cmd/ocsp-updater/main.go

```diff
@@ -107,97 +205,106 @@ func (updater *OCSPUpdater) processResponse(tx *gorp.Transaction, serial string)
+			updater.log.AuditErr(fmt.Errorf("Failed to store OCSP response: %s", err))
+			updater.stats.Inc("OCSP.Errors.StoreResponse", 1, 1.0)
+			tx.Rollback()
+			return
+		}
+	}
 	err = tx.Commit()
 	if err != nil {
-		updater.log.Err(fmt.Sprintf("OCSP %s: Error committing transaction, skipping: %s", serial, err))
-		updater.stats.Inc("OCSP.Updates.Failed", 1, 1.0)
-		tx.Rollback()
-		return err
+		updater.log.AuditErr(fmt.Errorf("Failed to commit OCSP response transaction: %s", err))
+		updater.stats.Inc("OCSP.Errors.CommitTransaction", 1, 1.0)
+		return
 	}
+	updater.stats.Inc("OCSP.StoredResponses", int64(len(responses)), 1.0)
 
-	updater.log.Info(fmt.Sprintf("OCSP %s: OK", serial))
-	updater.stats.Inc("OCSP.Updates.Processed", 1, 1.0)
-	updater.stats.TimingDuration("OCSP.Updates.UpdateLatency", time.Since(innerStart), 1.0)
-	return nil
+	return
 }
 
-// findStaleResponses opens a transaction and processes up to responseLimit
-// responses in a single batch. The responseLimit should be relatively small,
-// so as to limit the chance of the transaction failing due to concurrent
-// updates.
-func (updater *OCSPUpdater) findStaleResponses(oldestLastUpdatedTime time.Time, responseLimit int) error {
-	var certificateStatus []core.CertificateStatus
-	_, err := updater.dbMap.Select(&certificateStatus,
-		`SELECT cs.* FROM certificateStatus AS cs JOIN certificates AS cert ON cs.serial = cert.serial
-		 WHERE cs.ocspLastUpdated < ? AND cert.expires > now()
-		 ORDER BY cs.ocspLastUpdated ASC
-		 LIMIT ?`, oldestLastUpdatedTime, responseLimit)
+// oldOCSPResponsesTick looks for certificates with stale OCSP responses and
+// generates/stores new ones
+func (updater *OCSPUpdater) oldOCSPResponsesTick(batchSize int, now, _ time.Time) {
+	statuses, err := updater.findStaleOCSPResponses(now.Add(-updater.ocspMinTimeToExpiry), batchSize)
+	if err != nil {
+		updater.stats.Inc("OCSP.Errors.FindStaleResponses", 1, 1.0)
+		updater.log.AuditErr(fmt.Errorf("Failed to find stale OCSP responses: %s", err))
+		return
+	}
 
-	if err == sql.ErrNoRows {
-		updater.log.Info("All up to date. No OCSP responses needed.")
-	} else if err != nil {
-		updater.log.Err(fmt.Sprintf("Error loading certificate status: %s", err))
-	} else {
-		updater.log.Info(fmt.Sprintf("Processing OCSP Responses...\n"))
-		outerStart := time.Now()
-
-		for i, status := range certificateStatus {
-			updater.log.Debug(fmt.Sprintf("OCSP %s: (%d/%d)", status.Serial, i, responseLimit))
-
-			err = updater.updateOneSerial(status.Serial)
-			// Abort if we recieve a fatal error
-			if _, ok := err.(FatalError); ok {
-				return err
-			}
-		}
+	updater.generateOCSPResponses(statuses)
+}
 
-		updater.stats.TimingDuration("OCSP.Updates.BatchLatency", time.Since(outerStart), 1.0)
-		updater.stats.Inc("OCSP.Updates.BatchesProcessed", 1, 1.0)
-	}
+type looper struct {
+	clk       clock.Clock
+	stats     statsd.Statter
+	batchSize int
+	tickDur   time.Duration
+	tickFunc  func(int, time.Time, time.Time)
```

Comment:
Let's reverse the order of these args to be (earliest, latest), and document them.